src/framework/myCFG.ml

```diff
@@ -165,138 +165,214 @@ let createCFG (file: file) =
+            (* CIL uses empty Instr self-loop for empty Loop, so a Skip self-loop must be added to not lose the loop. *)
+            begin match real_succs () with
+              | [] -> () (* if stmt.succs is empty (which in other cases requires pseudo return), then it isn't a self-loop to add anyway *)
+              | [succ] ->
+                if CilType.Stmt.equal succ stmt then (* self-loop *)
+                  addEdge (Statement stmt) (Cil.locUnknown, Skip) (Statement succ) (* TODO: better loc from somewhere? *)
+              | _ -> failwith "MyCFG.createCFG: >1 Instr [] succ"
+            end
+
+          | Instr instrs -> (* non-empty Instr *)
+            let edge_of_instr = function
               | Set (lval,exp,loc) -> loc, Assign (lval, exp)
               | Call (lval,func,args,loc) -> loc, Proc (lval,func,args)
               | Asm (attr,tmpl,out,inp,regs,loc) -> loc, ASM (tmpl,out,inp)
               | VarDecl (v, loc) -> loc, VDecl(v)
             in
-            let handle_instrs' = function
-              (* Empty Instrs are weird: they have edges without any label or transfer function.
-               * Instead turn them into Skips, which keep them in Goblint's CFG for witness use. *)
-              | [] -> [Cil.locUnknown, Skip] (* TODO: better loc from somewhere? *)
-              | xs -> List.map handle_instr xs
-            in
-            let handle_instrs succ = mkEdges (Statement stmt) (handle_instrs' xs) succ in
-            (* Sometimes a statement might not have a successor.
-             * This can happen if the last statement of a function is a call to exit. *)
-            let succs = if stmt.succs = [] then [Lazy.force pseudo_return] else List.map (fun x -> Statement (realnode true x)) stmt.succs in
-            List.iter handle_instrs succs
-          (* If expressions are a bit more interesting, but CIL has done
-           * its job well and we just pick out the right successors *)
-          | If (exp, true_block, false_block, loc) -> begin
-              if isZero exp then ()
-              else
-                let false_stmt = realnode true (List.nth stmt.succs 0) in
-                let true_stmt = try
-                    realnode true (List.nth stmt.succs 1)
-                  with Failure _ -> realnode true (List.hd stmt.succs) in
-                mkEdge stmt (Test (exp, true )) true_stmt;
-                mkEdge stmt (Test (exp, false)) false_stmt
+            let edges = List.map edge_of_instr instrs in
+            let add_succ_node succ_node = addEdges (Statement stmt) edges succ_node in
+            begin match real_succs () with
+              | [] -> add_succ_node (Lazy.force pseudo_return) (* stmt.succs can be empty if last instruction calls non-returning function (e.g. exit), so pseudo return instead *)
+              | [succ] -> add_succ_node (Statement succ)
+              | _ -> failwith "MyCFG.createCFG: >1 non-empty Instr succ"
```

Comment:
Non empty? Or not empty nor singleton?

Reply:
There's two different things here: the list of `instr`s in `Instr` and the list of the successors of `Instr`. The "non-empty" is about the `Instr` argument while the `>1` is about the number of successors.

If my study of CIL's CFG preparation is correct, these failures should never happen, but I wanted to make them all really explicit, just in case. Previously they were just mapped over as lists, as if there could be arbitrarily many successors that duplicate the same instructions on the edge, which doesn't really make sense. And indeed it shouldn't really happen, but I want to have all of this code very clear about what CIL gives us and what it doesn't give us.